### ISSUE TITLE
Interface to let applications know that they are about to be checkpointed

### ISSUE THREAD

**Issue by adrianreber** [checkpoint-restore/criu#1427]
Recently I started to think about that it might be helpful to have an interface with which CRIU could tell an application that it will be checkpointed very soon. One use case that would benefit from such an interface would definitely be the JVM.

I have been told that if a JVM would know that it will be checkpointed it can do some cleanups of data which does not make sense to be checkpointed. Unused memory could be freed or secrets could be removed from the memory or devices (GPU, systemcfg file on ppc (#1425)) could be closed. Especially the closing of certain devices and files could be helpful for many applications.

With the existing action scripts it is already possible to set up something like such an interface, but maybe CRIU could define an interface which always works. Independent of action scripts. Application like the JVM could provide this interface if they want to. If it does not exist CRIU would just work as right now.

I do not really have an idea yet how such an interface could look like. A file, a pipe, a socket? Not sure. It would need to be an interface which CRIU can detect really fast. If it does not exist the checkpointing should not block or take longer.

The protocol between the application and CRIU should also be really simple, so that any application can implement if it wants to without any external dependency.

The same should be possible after restoring to let an application know that it just has been restored.

Does something like this maybe already exist in OpenVZ?

Any ideas how something like this could look like?

**Comment by avagin** [comment]
This interface has to be synchronous. I mean CRIU sends a signal and needs to get a replay when the application is ready.

>  Unused memory could be freed or secrets could be removed from the memory or devices (GPU, systemcfg file on ppc (#1425)) could be closed. 

This means that we need to have an opposite signal that will tell that C/R has been done and the app can continue running.

> I have been told that if a JVM would know that it will be checkpointed it can do some cleanups of data which does not make sense to be checkpointed. 

This can be useful for another use-case, when a host is going to be hibernated.

**Comment by Snorch** [comment]
> Does something like this maybe already exist in OpenVZ?

Sadly looks like there is no such thing in Virtuozzo criu...

> This means that we need to have an opposite signal that will tell that C/R has been done and the app can continue running.

If we have a synchronous interface it should also be _only_ available to criu, else any "virus"-app would be able to put Jave in this suspended state.

**Comment by brauner** [comment]
On Wed, May 19, 2021 at 05:03:21AM -0700, Pavel Tikhomirov wrote:
> > Does something like this maybe already exist in OpenVZ?
> 
> Sadly looks like there is no such thing in Virtuozzo criu...
> 
> > This means that we need to have an opposite signal that will tell that C/R has been done and the app can continue running.
> 
> If we have a synchronous interface it should also be _only_ available to criu, else any "virus"-app would be able to put Java in this suspended state.
> 
> upd: Though anyone who can ptrace (or freeze cgroup of) the task already can put it in suspended state... Maybe we can add some functionality to send c/r a request to the task via pidfd? @brauner 

Hm, you mean sending a signal via pidfd because that already works via
pidfd_send_signal() or are you talking about something more elaborate?

Christian

**Comment by Snorch** [comment]
>>Maybe we can add some functionality to send c/r a request to the task via pidfd?
>Hm, you mean sending a signal via pidfd because that already works via
pidfd_send_signal() or are you talking about something more elaborate?

I mean some kind of bidirectional interface which allows the task with pidfd for some process send command to this process and get reply when command was executed.. The thing that Andrew was talking about but through pidfd, because pidfd owner should be allowed to ask the process to "cleanup caches".

**Comment by nravic** [comment]
Is anyone working on this? I've been playing around a lot w/ CRIU and building a migration service - something like this would be very useful. Happy to take it on if no one else is :)

**Comment by xial-thu** [comment]
Any udpate?

I would propose a user-story: As a AI algorithm researcher, I've started a distributed training job and need to migrate a training process to another host. The processes will do all-reduce(sending data to other process, so that each process has all data), so I have to make sure there's no flying data when checkpointing happens(due to flying data cannot be dumped, to avoid the loss of data).

It would be nice if CRIU provides a machenism to let process know, so that they can do some necessary pre-processing.

**Comment by adrianreber** [comment]
> Any udpate?

No, nothing. I do not expect that anything will happen here soon. Sorry.

> I would propose a user-story: As a AI algorithm researcher, I've started a distributed training job and need to migrate a training process to another host. The processes will do all-reduce(sending data to other process, so that each process has all data), so I have to make sure there's no flying data when checkpointing happens(due to flying data cannot be dumped, to avoid the loss of data).

Thanks for the use case, but that is not the problem. The technical details are the problem.

But, regarding your use case, what do you mean by flying data. Can you be more specific. I am not sure it is a real problem, depending on how you use it.

> It would be nice if CRIU provides a machenism to let process know, so that they can do some necessary pre-processing.

Yes, it would be nice. You can always implement your own solution. You could `signal()` your process to get ready for checkpointing before running CRIU.

**Comment by nravic** [comment]
For what it's worth I've been playing around with this locally but haven't gotten very far since the holidays. Hoping to make some progress on it this month.

**Comment by nravic** [comment]
@adrianreber do you foresee any issues if I use something like `epoll` instead of directly wiring signals w/ pidfd_send_signal()?

**Comment by nravic** [comment]
here's a quick sketch: https://github.com/checkpoint-restore/criu/compare/criu-dev...nravic:criu:feat/signal_process 

would be great to know if it's on the right track. I'm thinking instead of bidirectional, it can just be a unidirectional signal w/ a timeout. So criu sends dumpee a signal and waits for a set timeout before continuing. 

It feels less brittle than waiting for a response back but at the same time it's very simple.

**Comment by rst0git** [comment]
@nravic A [`pre-dump`](https://criu.org/Action_scripts) action script could be used to achieve this as well.

**Comment by nravic** [comment]
@rst0git yeah good point - maybe I can write up a quick example for it for the docs instead. 

@adrianreber since you posited the initial issue, what do you think? Is there a benefit of having an interface for this type of behavior inside of criu that we don't get by using an action script?

**Comment by adrianreber** [comment]
Your patch looks like a reasonable implementation, but something which I would not expect to be merged in CRIU. It seems like it only works for your setup as there is no standardized interface and the waiting is also something which doesn't sound right. The signalling should not slow down checkpointing.

I think an action script is the best solution. It is independent of CRIU and you can adapt it for your use case without changing CRIU's source code and recompilation.

**Comment by nravic** [comment]
Yeah agreed. I don't really have use for a standardized interface now anyway and I'm not sure the juice is worth the squeeze. 

Thanks for looking into it!
